crates/ruff_python_formatter/src/expression/expr_tuple.rs

```diff
@@ -1,19 +1,156 @@
+use crate::context::{NodeLevel, PyFormatContext};
 use crate::expression::parentheses::{
     default_expression_needs_parentheses, NeedsParentheses, Parentheses, Parenthesize,
 };
-use crate::{not_yet_implemented_custom_text, FormatNodeRule, PyFormatter};
-use ruff_formatter::{write, Buffer, FormatResult};
+use crate::trivia::find_first_non_trivia_character_after;
+use crate::{AsFormat, FormatNodeRule, PyFormatter, USE_MAGIC_TRAILING_COMMA};
+use ruff_formatter::formatter::Formatter;
+use ruff_formatter::prelude::{
+    block_indent, group, hard_line_break, if_group_breaks, soft_block_indent, soft_line_break,
+    soft_line_break_or_space, text,
+};
+use ruff_formatter::{format_args, write, Buffer, Format, FormatResult};
+use ruff_python_ast::prelude::{Expr, Ranged};
+use ruff_text_size::{TextLen, TextRange};
 use rustpython_parser::ast::ExprTuple;
 
 #[derive(Default)]
 pub struct FormatExprTuple;
 
 impl FormatNodeRule<ExprTuple> for FormatExprTuple {
-    fn fmt_fields(&self, _item: &ExprTuple, f: &mut PyFormatter) -> FormatResult<()> {
-        write!(f, [not_yet_implemented_custom_text("(1, 2)")])
+    fn fmt_fields(&self, item: &ExprTuple, f: &mut PyFormatter) -> FormatResult<()> {
+        let ExprTuple {
+            range,
+            elts,
+            ctx: _,
+        } = item;
+
+        // Handle the edge cases of an empty tuple and a tuple with one element
+        let last = match &elts[..] {
+            [] => {
+                return text("()").fmt(f);
```

Comment:
Can you add a test for an empty tuple that contains a comment: 

```python
(
	# empty
)
```

You'll need to handle that case manually by:

* overriding `fmt_dangling_comments` on `FormatNodeRule`
* Use `block_indent(&dangling_node_comments(item.into())` here to format the dangling comments proper

Reply:
didn't realize earlier, but it's neat that block_ident is a noop when empty